spotify-api.el

```diff
@@ -152,67 +155,89 @@ that runs a local httpd for code -> token exchange."
+				(spotify-persist-token token now)
+        (if (null token)
+          (user-error "OAuth2 authentication failed")
+          token))
+			;; Spotify tokens appear to expire in 3600 seconds (60 min). We renew
+			;; at 3000 (50 min) to play it safe
+      (if (> now (+ *spotify-oauth2-ts* 3000))
+        (let ((token (oauth2-refresh-access *spotify-oauth2-token*)))
+					(spotify-persist-token token now)
           (if (null token)
-              (user-error "OAuth2 authentication failed")
+            (user-error "Could not refresh OAuth2 token")
             token))
-      (if (> now (+ *spotify-oauth2-ts* 3000))
-          (let ((token (oauth2-refresh-access *spotify-oauth2-token*)))
-            (setq *spotify-oauth2-token* token)
-            (setq *spotify-oauth2-ts* now)
-            (if (null token)
-                (user-error "Could not refresh OAuth2 token")
-              token))
         *spotify-oauth2-token*))))
 
-(defun spotify-api-call-async (method uri &optional data callback is-retry)
+(defun spotify-api-call-async (method uri &optional data callback)
   "Make a request to the given Spotify service endpoint URI via METHOD.
 Call CALLBACK with the parsed JSON response."
-  (oauth2-url-retrieve
-   (spotify-oauth2-token)
-   (concat spotify-api-endpoint uri)
-   (lambda (_)
-     (toggle-enable-multibyte-characters t)
-     (goto-char (point-min))
-     (condition-case _
-         (when (search-forward-regexp "^$" nil t)
-           (let* ((json-object-type 'hash-table)
-                  (json-array-type 'list)
-                  (json-key-type 'symbol)
-                  (json (json-read))
-                  (error-json (gethash 'error json)))
-             (kill-buffer)
-
-             ;; Retries the request when the token expires and gets refreshed
-             (if (and (hash-table-p error-json)
-                      (eq 401 (gethash 'status error-json))
-                      (not is-retry))
-                 (spotify-api-call-async method uri data callback t)
-               (when callback (funcall callback json)))))
-
-       ;; Handle empty responses
-       (end-of-file
-        (kill-buffer)
-        (when callback (funcall callback nil)))))
-   nil
-   method
-   (or data "")
-   '(("Content-Type" . "application/json"))))
+	(request (concat spotify-api-endpoint uri)
+		:headers `(("Authorization" .
+								 ,(format "Bearer %s" (oauth2-token-access-token (spotify-oauth2-token))))
+								("Accept" . "application/json")
+								("Content-Type" . "application/json")
+								("Content-Length" . ,(length data)))
+		:type method
+		:parser (lambda ()
+							(let ((json-object-type 'hash-table)
+										 (json-array-type 'list)
+										 (json-key-type 'symbol))
+								(when (> (buffer-size) 0)
+									(json-read))))
+		:encoding 'utf-8
+		:data data
+		:success (cl-function
+							 (lambda (&key response &allow-other-keys)
+								 (when callback (funcall callback (request-response-data response)))))))
```

Comment:
This is a much better API to work with, it seems 😌

Reply:
So much less stress in my world! And it's been performing admirably since I put it in locally.

Reply:
I'm testing it right now and it's working nicely indeed.